lib/pwned_password_downloader.rb

```diff
@@ -0,0 +1,112 @@
+#!/usr/bin/env ruby
+require 'fileutils'
+require 'net/http/persistent'
+require 'retries'
+require 'ruby-progressbar'
+require 'ruby-progressbar/outputs/null'
+
+class PwnedPasswordDownloader
+  attr_reader :destination,
+              :num_threads,
+              :keep_threshold,
+              :output_progress
+
+  alias_method :output_progress?, :output_progress
+
+  RANGE_API_ROOT = 'https://api.pwnedpasswords.com/range/'
+  SHA1_LENGTH = 40
+  HASH_PREFIX_LENGTH = 5
+  OCCURRENCE_OFFSET = SHA1_LENGTH - HASH_PREFIX_LENGTH + ':'.length
+
+  def initialize(
+    destination: 'tmp/pwned',
+    num_threads: 64,
+    keep_threshold: 30,
+    output_progress: true
+  )
+    @destination = destination
+    @num_threads = num_threads
+    @keep_threshold = keep_threshold
+    @output_progress = output_progress
+  end
+
+  def run!(start: '00000', finish: 'FFFFF')
+    (start.to_i(16)..finish.to_i(16)).each do |prefix_num|
+      queue << prefix_num.to_s(16).upcase.rjust(5, '0')
+    end
+
+    FileUtils.mkdir_p(destination)
+
+    progress_bar = ProgressBar.create(
+      title: 'Downloading...',
+      total: queue.size,
+      output: output_progress? ? $stdout : ProgressBar::Outputs::Null,
+      format: '[ %t ] %p%% %B %a (%E)',
+    )
+
+    [num_threads, queue.size].min.times.map do
+      Thread.new do |thread_id|
+        net_http = Net::HTTP::Persistent.new(name: "thread_id_#{thread_id}")
+
+        while (prefix = queue.pop)
+          if already_downloaded?(prefix)
+            progress_bar.increment
+            next
+          end
+
+          begin
+            write_one(
+              prefix:,
```

Comment:
In a previous meeting, I recall a decision to avoid this shorthand keyword/hash syntax in Ruby. I've noticed it has crept into our code in a variety of locations.

Reply:
From what I can recall, previous discussions were about whether to enforce one way or the other. There hasn't been agreement on this, so my understanding is that it's left to the developer's discretion.